llama-index-core/llama_index/core/query_pipeline/query.py

```diff
@@ -365,8 +393,36 @@ def run_multi(
                 CBEventType.QUERY,
                 payload={EventPayload.QUERY_STR: json.dumps(module_input_dict)},
             ) as query_event:
-                outputs, _ = self._run_multi(module_input_dict)
-                return outputs
+                if batch:
+                    outputs = {}
+
+                    batch_size = min(
+                        len(values)
+                        for subdict in module_input_dict.values()
+                        for values in subdict.values()
+                    )
+
+                    # List individual outputs from batch multi input.
+                    inputs = [
+                        {
+                            key: {
+                                inner_key: inner_val[i]
+                                for inner_key, inner_val in value.items()
+                            }
+                            for key, value in module_input_dict.items()
+                        }
+                        for i in range(batch_size)
+                    ]
+                    for input in inputs:
```

Comment:
So, one of the ideas behind batch mode is being able to process them all at the same time.

This can be done with threading/multiprocessing, but tbh, thats typically more complicated to get working than its worth.

Another idea would be to use async to run just concurrently (assuming components aren't using fake-async)

So in both the `run()` and `arun()` methods, we can use async to just run everything.

in `arun()`
```
jobs = [self._arun_multi(input) for input in inputs]
results = await asyncio.gather(*jobs)
```

in run()
```
jobs = [self._arun_multi(input) for input in inputs]
results = asyncio.run(asyncio.gather(*jobs))
```

Reply:
Oh ok, thanks!

Reply:
Let me fix this up rq!